engines/cryo/cryolib.cpp

```diff
@@ -0,0 +1,453 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include "common/system.h"
+#include "common/events.h"
+#include "common/timer.h"
+
+#include "graphics/palette.h"
+
+#include "cryo/cryo.h"
+#include "cryo/cryolib.h"
+
+namespace Cryo {
+
+///// Mac APIs
+
+void SysBeep(int x) {
+}
+
+void FlushEvents(int16 arg1, int16 arg2) {
+}
+
+///// CLView
+
+View::View(int w, int h) {
+	void *buffer = (byte *)malloc(w * h);
+	if (buffer)
+		initDatas(w, h, buffer);
+	else
+		error("Unable to allocate view buffer");
+}
+
+View::~View() {
+	if (_bufferPtr)
+		free(_bufferPtr);
+}
+
+// Original name: CLView_SetSrcZoomValues
+void View::setSrcZoomValues(int x, int y) {
+	_zoom._srcLeft = x;
+	_zoom._srcTop = y;
+}
+
+// Original name: CLView_SetDisplayZoomValues
+void View::setDisplayZoomValues(int w, int h) {
+	_zoom._width = w;
+	_zoom._height = h;
+}
+
+// Original name: CLView_InitDatas
+void View::initDatas(int w, int h, void *buffer) {
+	_bufferPtr = (byte *)buffer;
+	_width = w;
+	_height = h;
+	_pitch = w;
+	_doubled = false;
+	_normal._srcLeft = 0;
+	_normal._srcTop = 0;
+	_normal._dstLeft = 0;
+	_normal._dstTop = 0;
+	_normal._width = w;
+	_normal._height = h;
+	_zoom._srcLeft = 0;
+	_zoom._srcTop = 0;
+	_zoom._dstLeft = 0;
+	_zoom._dstTop = 0;
+	_zoom._width = w;
+	_zoom._height = h;
+}
+
+// Original name: CLView_CenterIn
+void View::centerIn(View *parent) {
+	_normal._dstLeft = (parent->_width - _normal._width) / 2;
+	_normal._dstTop = (parent->_height - _normal._height) / 2;
+	_zoom._dstLeft = (parent->_width - _zoom._width) / 2;
+	_zoom._dstTop = (parent->_height - _zoom._height) / 2;
+}
+
+///// CLPalette
+uint16 gIntervalLast, gIntervalFirst, gIntervalSet;
+int16 gMacintize = 0;
+color_t black_palette[256];
+color_t last_palette[256];
+
+void CLPalette_Init() {
+	for (int16 i = 0; i < 256; i++)
+		black_palette[i].r = black_palette[i].g = black_palette[i].b = 0;
+}
+
+void CLPalette_SetLastPalette(color_t *palette, int16 first, int16 count) {
+	for (int16 i = first; i < first + count; i++)
+		last_palette[i] = palette[i];
+}
+
+void CLPalette_GetLastPalette(color_t *palette) {
+	for (int16 i = 0; i < 256; i++)
+		palette[i] = last_palette[i];
+}
+
+void CLPalette_SetRGBColor(color_t *palette, uint16 index, color3_t *rgb) {
+	palette[index].r = rgb->r;
+	palette[index].g = rgb->g;
+	palette[index].b = rgb->b;
+	palette[index].a = 0;
+}
+
+void CLPalette_Macintize(int16 macintize) {
+	gMacintize = macintize;
+}
+
+void CLPalette_SetInterval(uint16 first, uint16 last) {
+	gIntervalFirst = first;
+	gIntervalSet = 1;
+	gIntervalLast = last;
+}
+
+void CLPalette_DeactivateInterval() {
+	gIntervalSet = 0;
+}
+
+void CLPalette_Send2Screen(struct color_t *palette, uint16 first, uint16 count) {
+	if (gMacintize) {
+		palette[0].r = palette[0].g = palette[0].b = 0xFFFF;
```

Comment:
Again, this looks odd... shouldn't the r/g/b components be bytes, instead of 16-bit words?